include/net/rest_client.h

```diff
@@ -86,6 +86,7 @@ struct rest_client_req_context {
 	/** User-defined timeout value for REST request. The timeout is set individually
 	 *  for socket connection creation and data transfer meaning REST request can take
 	 *  longer than this given timeout. To disable, set the timeout duration to SYS_FOREVER_MS.
+	 *  A value of zero will result in an immediate timeout.
```

Comment:
Isn't that backwards?  I thought the change was making 0 wait forever?

Reply:
this comment specifies the behavior of `rest_client`.
`nrf_cloud_rest` is handling zero as a no-timeout (wait forever) value, since it doesn't make sense to immediately timeout.

Reply:
Does it make sense for `rest_client` to immediately timeout?

Reply:
🤷‍♂️ 
i don't think so, but that is how it is implemented.
@trantanen  is a reviewer, so he can comment about `rest_client`.

Reply:
I guess it depends on if we want rest_client to remain conformal with how Zephyr, in general, handles timeouts. Only `SYS_FOREVER_MS` should really ever be used to represent indefinite timeouts. 

To me, maintaining that consistency, especially in lower-level libraries like rest_client, is valuable.

I'm on the fence about whether having a zero-check in nrf_cloud_rest is alright. I might even argue that instead of adding the zero-check, as this PR does, we should update all of the users of the library to use `SYS_FOREVER_MS`.

There are a couple spots -- I think the POSIX sockets themselves -- which treat 0 as an indefinite timeout, so it's a bit odd that Zephyr has this `SYS_FOREVER_MS` convention, but that is the convention that is in place.

Reply:
Originally we chose to have zero as indefinite timeout in Location library without knowing what Zephyr's convention was. @glarsennordic wrote a note about that in PR#7978 after which we changed Location library and some related places like rest_client:
https://github.com/nrfconnect/sdk-nrf/pull/7978#pullrequestreview-1015713855

The reason I really like zero as indefinite timeout is that when you initialize a context/config structure, such as 
`nrf_cloud_rest_context`, `rest_client_req_context`, `multicell_location_params`, or `location_gnss_config` to zero, timeout is disabled without further actions such as setting timeout to `SYS_FOREVER_MS` etc. But I don't know if we want to discuss going back to zero indefinite timeout instead of `SYS_FOREVER_MS`.

Zero handling hasn't been mentioned intentionally in the APIs that I modified because various components treat it differently so the behavior is undefined. Some handle it as indefinite timeout and some with instant timeout mostly depending on how things happen in the way the code  and input structures are structured and set to defaults. I didn't like to make `SYS_FOREVER_MS` and zero as indefinite timeouts as we were aligning with Zephyr's convention. Otherwise we could have just left zero be indefinite timeout and ignore `SYS_FOREVER_MS`. I don't know if this decision was good but that's what I thought at the time. I still think we shouldn't have those two values defined as indefinite timeout.

> I might even argue that instead of adding the zero-check, as this PR does, we should update all of the users of the library to use SYS_FOREVER_MS.

Unless we open the discussion about using zero as indefinite timeout throughout nrf libraries, I would probably continue the way I chose, i.e., change all nrf_cloud_rest related code to use `SYS_FOREVER_MS` for indefinite timeout. If nrf_cloud_rest library wants to handle zero as indefinite timeout, it could do so without specifying it.

Reply:
```
The reason I really like zero as indefinite timeout is that when you initialize a context/config structure, such as
nrf_cloud_rest_context, rest_client_req_context, multicell_location_params, or location_gnss_config to zero, timeout is disabled without further actions such as setting timeout to SYS_FOREVER_MS etc. But I don't know if we want to discuss going back to zero indefinite timeout instead of SYS_FOREVER_MS.
```

Yes I agree, zero is generally a default value and it is more user friendly for `nrf_cloud_rest` to have it act as a no-timeout (indefinite) value.

Correct, this PR is not to re-discuss how the timeout values could be handled.  It is to fix the broken `nrf_cloud_rest` library/samples.

Reply:
`Zero handling hasn't been mentioned intentionally in the APIs that I modified because various components treat it differently so the behavior is undefined.`

the immediate timeout occurs here:
https://github.com/nrfconnect/sdk-nrf/blob/2797a7acb70d8f6d7b2fcc6391b3dabaf45d11b7/subsys/net/lib/rest_client/src/rest_client.c#L313
even if k_uptime() returns the same value between calls, a zero value for `timeout_ms` will result in a timeout.
so i believe the comment i added to `rest_client.h` is necessary.

Reply:
I know immediate timeout occurs there. However, that doesn't mean we need to state that in the description for zero although you seem to prefer to write it down.
@glarsennordic what do you think we should do for zero handling?

Reply:
Sorry, I missed your ping @trantanen

Personally, I think an error should be raised if zero or negative is passed, and we should otherwise stick to the Zephyr convention (only SYS_TIMEOUT_FOREVER should be indefinite timeout)

But I don't see it as the end of the world if we decide to interpret zero as indefinite timeout instead.

However, whatever we choose to do, I think we should fully document it. For me, the appeal of sticking to the Zephyr convention is that it reduces confusion -- if we deviate from it, we need to document our deviation to minimize the confusion that deviation may create.

Reply:
Yep, seems this got merged. However, right now we are not particularly consistent in terms of functionality (zero handling differs between libraries) or documentation part.